### PR TITLE
feat(api): delete the authenticated user

### LINKED ISSUE #19439: [API] Delete Yourself
the user should have an api to delete his own account

you just have to look at
https://github.com/go-gitea/gitea/blob/a7f0ce620774be861f11b57d426ab59e27e4171a/routers/api/v1/admin/user.go#L307-L320

### PR DESCRIPTION
**Pull request by dhruvmanila**
Delete the authenticated user.

Endpoint: `DELETE /user`

resolves: #19439

### CODE REVIEW

File: routers/api/v1/user/user.go
Comment (6543) [inline routers/api/v1/user/user.go:160]: Please document purge

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
thanks

**Comment by wxiaoguang** [comment]
I do not have a good feeling about this either. What's the real use case / real user story for such API?

**Comment by dhruvmanila** [comment]
Hey @zeripath @wxiaoguang, I get where you're coming from and the problems this API might create. As per the discussions in the linked issue (#19439), it seemed clear that for this API there's no point in having the "confirmation" part. That part should be taken care by the frontend.

As per my limited knowledge of `gitea` (this is my first PR), I think before this only an admin can delete a user but given this API, a frontend component could be added to delete yourself and the frontend part will take care of the confirmation part.

**Comment by wxiaoguang** [comment]
Usually, the web backend and api backend is different. And there were some PRs to decouple the API endpoints from the web. Some backgrounds:
* https://github.com/go-gitea/gitea/pull/16052
* https://github.com/go-gitea/gitea/pull/19318
* https://github.com/go-gitea/gitea/pull/19321

If a function is only used by web UI, then it's not necessary to make it a common API.

**Review by 6543 — APPROVED** [review]

File: routers/api/v1/api.go
Comment (6543) [inline routers/api/v1/api.go]: require basicAuth

**Review by 6543 — CHANGES_REQUESTED** [review]

**Comment by 6543** [comment]
> This API could be a serious problem.

agree but:
 - for acidential -> the api client is responsible for that
 - for malicious: https://github.com/go-gitea/gitea/pull/20410#discussion_r980636870, so on app with just a token can dell stuff (similar to how the token endpoint's are handled)

File: routers/api/v1/api.go
Comment (6543) [reply routers/api/v1/api.go]: @dhruvmanila just use `reqBasicOrRevProxyAuth()` helper

**Comment by wxiaoguang** [comment]
I have strong objection about making it as a `/api/` API before there is a clear design, quote comment https://github.com/go-gitea/gitea/issues/19439#issuecomment-1197636298

> So, if a user can only delete themself from the web UI, then function can be written in the web backend code.

> If a user can delete themself by API, then I think there need more details: why they need to do so, what's the real use case?

**Comment by dhruvmanila** [comment]
Oh sorry, will keep that in mind from next time 👍

**Comment by hello-smile6** [comment]
> Oh sorry, will keep that in mind from next time 👍

Is this ready?

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
did add an smal code-format to make it more redable

**Review by wxiaoguang — DISMISSED** [review]
~~I still have the question why it should be an API, what's the real use case, why a user would call an API to delete themself.~~

outdated

**Comment by wxiaoguang** [comment]
Sorry for bothering, I didn't say clearly before. To summary my opinion:

Concepts: 

* Public API: `/api/xxx`, could only be called with tokens, can not be called by Gitea web frontend with session at the moment. They are preferred to be stable and prevented from breakings.
* Web API: the normal `/user/xxx`, could only be called by Gitea web frontend with sessions, can be refactored easily.

Opinions:

1. The Web API and Public API are different. If a function is only used by web UI, then it's not necessary to make it a Public API.
2. Gitea itself needs the "Delete Account" feature too, on the web.
3. For various native clients, they could redirect users to the "Delete Account" web page to finish the operations. As a destructive feature, it's not suitable to be publicly called.
4. At the moment, a Public API can not be called from Gitea web frontend. So to satisfy (2), the API should be put in the web backend routes as Web API.
5. Since this API is destructive, maybe it should be improved: should the user re-input their password and 2FA to delete their account? A single API call without check seems pretty dangerous, if there is something like token lost or XSS.

**Comment by wxiaoguang** [comment]
Hmm, just saw the latest `reqBasicOrRevProxyAuth` change. It should be fine? I will dismiss my request change and hide out-dated comments.

File: routers/api/v1/user/user.go
Comment (lunny) [inline routers/api/v1/user/user.go:182]: We should ensure only site administrator could do it.

**Comment by lunny** [comment]
Please move the api under `/api/v1/admin/users`.

**Comment by wxiaoguang** [comment]
When this PR mixes with 
* https://github.com/go-gitea/gitea/pull/15119#issuecomment-1261916745

When the instance is running behind reverse-proxy auth, other users with access  token (without password/otp auth) can delete the token's user.

**Comment by denyskon** [comment]
What is the state of this PR? Are you still working on it?

**Comment by wxiaoguang** [comment]
> What is the state of this PR? Are you still working on it?

My opinion is:

1. I do not think Gitea has a good and safe auth system design. That's the root problem.
2. The account deletion should have extra confirmation, but some features made the situation more complex (eg: the reverse-proxy header auth).
3. The account deletion API doesn't seem bring enough benefit but it might cause security & design problem if some cases are missing.


I would suggest to clarify the details before continuing. eg: various edge cases, what's the correct approach for auth and account deletion.

----

That's just my opinion, if most maintainers think this is right, I won't block.

**Comment by denyskon** [comment]
I think I'd agree with you on this. If it brings security risks, I don't think this feature is that much of a benefit that we could ignore this issue. Besides the fact that I don't even really see a reason for such an API endpoint 😆

**Comment by dhruvmanila** [comment]
> What is the state of this PR? Are you still working on it?

Hey, no I'm currently not working on this. But from all the discussion around this, I think it might be useful to take a step back and see if this is actually beneficial. It's been a while so I've lost context around this 😅 

That said, I'm totally fine in closing the PR for the time being and someone can pick it up if we decide to go forward.

**Comment by denyskon** [comment]
Okay, so let's close this and move the discussion back to the issue

**Pull request closed without merging**
